### DISCUSSION TITLE
Can't generate some Setup API functions for AnyCPU

### DISCUSSION THREAD

**Discussion by Slion** [microsoft/CsWin32#592] [Q&A] [answered]
I'm getting that warning/error:

`warning PInvoke005: This API is only available when targeting a specific CPU architecture. AnyCPU cannot generate this API.`

For the following functions:

```
SetupDiEnumDeviceInfo
SetupDiEnumDeviceInterfaces
SetupDiGetDeviceInterfaceDetail
SetupDiGetDevicePropertyKeys
SetupDiGetDeviceProperty
SetupDiGetDeviceInstanceId
```

Could this be fixed or are they really good reasons for keeping it that way?

**Comment by AArnott** [comment, marked as answer]
The C# representation for such APIs actually varies across architectures. Win32 took opportunity to change some structs and even some functions when switching to 64-bit in a way that C# cannot represent both versions with the same syntax. That's why CsWin32 cannot generate them unless you commit your compilation to one or more architecture-specific platforms. You can target x86, x64, and arm64 (compiling 3x with one project) and CsWin32 will generate the appropriate APIs during each compilation.

#409 and #366 describe possible ways to make this a bit better.
